[508] Form array buttons need unique aria-labels

Mottie opened the pull request:
## Description

Form array cards which don't include an easy method to make unique title and `aria-labels` on buttons. The title has a `includeIndexInTitle` ui option which adds a one-based index after the title, and the edit button has a `ariaLabelForEditButtonOnReview` ui option, but only for the review page. But there isn't any way to uniquely distinguish card buttons while editing.

This PR adds a `itemAriaLabel` ui option which allows you to convert the item data (from the array) and return a unique label to build upon.

```js
'ui:options': {
  itemAriaLabel: data => data[keyInformation],
}
```

Using this text, the `aria-label` for the array card edit, update, save, cancel and remove buttons are created so the screenreader user is able to get the full details on what the button will do (see the screenshot).

Thie PR also updates form 526 add disabilities, military history, POW and VA medical records pages which all use this data pattern.

**Note** If the array data includes an object, then the `ObjectField` will also need a duplicate `ui:options` set within the `items` section of the `uiSchema`.

Related:
- Ticket: https://github.com/department-of-veterans-affairs/va.gov-team/issues/18923
- 526 custom `ArrayField` used `itemKeyForAriaLabel`, but it has since been removed and `itemAriaLabel` from this PR will replace iit - https://github.com/department-of-veterans-affairs/vets-website/pull/16098

## Testing done

Updated unit tests

## Screenshots

<details><summary>Medical records</summary>

<!-- leave a blank line above -->
![Screen Shot 2021-02-19 at 9 26 57 AM](https://user-images.githubusercontent.com/136959/108525533-ca9e5600-7295-11eb-89da-f2bd0fb67b5f.png)</details>

<details><summary>Service history</summary>

<!-- leave a blank line above -->
![Screen Shot 2021-02-18 at 11 23 47 AM](https://user-images.githubusercontent.com/136959/108396606-a2065580-71dc-11eb-9838-388e8b9010ec.png)
![Screen Shot 2021-02-18 at 11 34 18 AM](https://user-images.githubusercontent.com/136959/108397617-c1ea4900-71dd-11eb-94f1-9411e8f8b7f6.png)

Update:
![](https://user-images.githubusercontent.com/136959/108732530-46e1a500-74f3-11eb-9b7b-162afb3cc264.png)</details>

<details>
<summary>POW periods</summary>

<!-- leave a blank line above -->
![Screen Shot 2021-02-19 at 2 05 42 PM](https://user-images.githubusercontent.com/136959/108556508-b7eb4780-72bc-11eb-93fa-15c614f20520.png)</details>

<details><summary>New conditions (review & submit page)</summary>

<!-- leave a blank line above -->
![Screen Shot 2021-02-18 at 12 15 35 PM](https://user-images.githubusercontent.com/136959/108403182-0f69b480-71e4-11eb-8cc4-0d0f0a3615cb.png)</details>



## Acceptance criteria
- [x] Array data cards include unique `aria-labels` on the card buttons on the chapter page
- [x] Array data cards include unique `aria-labels` on the card buttons on the review & submit page

## Definition of done
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [x] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [x] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

Mottie committed:
Make review & submit button aria-labels unique

Mottie committed:
Add unit tests

Mottie committed:
Add duplicate ui option for ObjectField

Mottie committed:
Remove extra text

Mottie committed:
Change to itemAriaLabel function

Mottie committed:
Fix labels on other parts of the form

micahchiang commented on src/applications/disability-benefits/all-claims/pages/militaryHistory.js:
```diff
@@ -58,7 +58,7 @@ export const uiSchema = {
       'ui:options': {
         itemName: 'Service Period',
-        itemKeyForAriaLabel: 'serviceBranch',
+        itemAriaLabel: data => data.serviceBranch,
```
Not sure how big a deal this is or how common, but if service branches are the same in two separate service periods, this would create identical aria labels. Would it be good to add some type of unique identifier per label? 
<img width="1188" alt="Screen Shot 2021-02-19 at 4 10 07 PM" src="https://user-images.githubusercontent.com/15097156/108562022-37811280-72cd-11eb-8677-943c35fe29b0.png">

Mottie replied:
Yeah, I could include the start date:

```js
itemAriaLabel: data => `${data.serviceBranch} starting on ${data.dateRange.from}`
```

Mottie replied:
Hmmm... so I included the date range, but I also have that same label used in the edit title.. not sure which is better "Editing service period" or this longer version

![Screen Shot 2021-02-22 at 9 37 57 AM](https://user-images.githubusercontent.com/136959/108732530-46e1a500-74f3-11eb-9b7b-162afb3cc264.png)

micahchiang replied:
Yeah I see what you mean, that is quite verbose. Would it be sufficient to shorten it to what you had suggested above?
``` 
itemAriaLabel: data => `${data.serviceBranch} starting on ${formatDate(data.dateRange.from)} 
```

micahchiang replied:
Totally your call though. Feel like this is becoming a content / wording discussion, and I don't want it to be a blocker from merging the PR 🤷

Mottie replied:
LOL, I might just go back to "Editing service branch", but I'll see what other reviewers say, since this will be a global change.

Mottie replied:
I also thought about passing all array data to the new function, to determine if there are duplicates, and if there are get the index and add "first", "second", etc - using https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/platform/forms-system/src/js/utilities/data/numberToWords.js

Mottie committed:
Include date range in label/title

bkjohnson commented on src/platform/forms-system/src/js/fields/ArrayField.jsx:
```diff
@@ -267,10 +265,15 @@ export default class ArrayField extends React.Component {
             const updateText = showSave && index === 0 ? 'Save' : 'Update';
             const isLast = items.length === index + 1;
             const isEditing = this.state.editing[index];
+            const ariaLabel = uiOptions.itemAriaLabel;
+            const itemName =
+              (typeof ariaLabel === 'function' && ariaLabel(item || {})) ||
+              uiOptions.itemName ||
+              'Item';
```
For `ArrayField`, `itemName` will always be _something_, even if it is "Item".

bkjohnson commented on src/platform/forms-system/src/js/review/ArrayField.jsx:
```diff
@@ -240,6 +240,11 @@ class ArrayField extends React.Component {
               (!schema.minItems || items.length > schema.minItems);
             const itemSchema = this.getItemSchema(index);
             const itemTitle = itemSchema ? itemSchema.title : '';
+            const ariaLabel = uiOptions.itemAriaLabel;
+            const itemName =
+              (typeof ariaLabel === 'function' && ariaLabel(item || {})) ||
+              uiOptions.itemName ||
+              'Item';
```
Same story here - `itemName` will be "Item" if it can't be anything else.

bkjohnson commented on src/platform/forms-system/src/js/review/ObjectField.jsx:
```diff
@@ -141,8 +141,14 @@ class ObjectField extends React.Component {
         title = title(formData, formContext);
       }
+      const uiOptions = uiSchema['ui:options'] || {};
+      const ariaLabel = uiOptions.itemAriaLabel;
+      const itemName =
+        (typeof ariaLabel === 'function' && ariaLabel(formData || {})) ||
+        formData[uiOptions.itemName];
       const editLabel =
-        _.get('ui:options.ariaLabelForEditButtonOnReview', uiSchema) ||
+        (itemName && `Edit ${itemName}`) ||
+        uiOptions.ariaLabelForEditButtonOnReview ||
         `Edit ${title}`;
```
But here, `itemName` can be null, which means we default to `ariaLabelForEditButtonOnReview` or the `title`. Should the label pattern here match what is happening in `ArrayField`?

Mottie replied:
Ahhh good catch, thanks! I'll add this.

Mottie replied:
Hmm, actually when I added a fallback to `'Item'`, it ends up that the `editLabel` always defaults to `Edit Item` because most object fields (not array fields) use the `title` on the review & submit page.

I added a duplicate `itemAriaLabel` to the ui options because the Military History section, new condition followup questions and VA treatment center name is rendered by the `ObjectField` on the review & submit page. Without the `itemAriaLabel` it would render a duplicate `'Edit {title}` for every entry.

![Screen Shot 2021-02-22 at 2 07 01 PM](https://user-images.githubusercontent.com/136959/108763773-5f63b680-7517-11eb-9e80-8bf88a71b5ab.png)

bkjohnson replied:
I'm not sure what I'm supposed to be seeing in the screenshot. Is this a problem or is this ideal?

Mottie replied:
We want the `aria-label` of the edit button to let a screenreader user know what the edit button will do. But as Micah pointed out, if the Veteran has more than one service period within the same branch, maybe we need to add a start date as well

bkjohnson commented on src/platform/forms-system/src/js/review/ArrayField.jsx:
```diff
@@ -284,13 +289,20 @@ class ArrayField extends React.Component {
                           <div className="small-6 left columns">
-                            <button className="float-left">Update</button>
+                            <button
+                              type="button"
```
Adding `type="button"` seems like a good call :+1:

Mottie replied:
Grrr actually, adding `type="button"` to this update button breaks submission since it doesn't have an `onclick` handler. I'm gonna set it to `type="submit"` to be more explicit

bkjohnson commented on src/platform/forms-system/src/js/fields/ArrayField.jsx:
```diff
@@ -351,9 +352,10 @@ export default class ArrayField extends React.Component {
                     className="usa-button-secondary edit vads-u-flex--auto"
+                    aria-label={`Edit ${itemName}`}
                     onClick={() => this.handleEdit(index)}
-                    aria-label={`Edit ${title}`}
```
Would it be better if we had `itemName` default to `title` instead of "Item" for the Edit and Update buttons?

Mottie replied:
I'm not sure if that is a good idea because the page title doesn't always match the item... e.g. for the prisoner of war page confinement ranges, the title is undefined, but we do define an item name.

bkjohnson reviewed (commented):
Some questions/observations about the differences between the `ArrayField`s and `ObjectField`.

Mottie committed:
Fix military history output

Mottie committed:
Add fallback to "Item" in ObjectField

Mottie committed:
Add more ObjectField labels

bkjohnson reviewed (commented):
I don't want this to creep out of scope, but this may help us reason better about the label in `ObjectField`. Here is a diff where I've removed all instances of `ariaLabelForEditButtonOnReview`. Let me know what you think:

<details>
    <summary>Diff</summary>

```diff
    diff --git a/src/applications/disability-benefits/996/pages/contestedIssues.js b/src/applications/disability-benefits/996/pages/contestedIssues.js
index 9ca782124d..7dd7351ee5 100644
--- a/src/applications/disability-benefits/996/pages/contestedIssues.js
+++ b/src/applications/disability-benefits/996/pages/contestedIssues.js
@@ -12,7 +12,7 @@ const contestedIssuesPage = {
   uiSchema: {
     'ui:title': ContestedIssuesTitle,
     'ui:options': {
-      ariaLabelForEditButtonOnReview: 'Edit issues eligible for review',
+      itemName: 'issues eligible for review',
     },
     contestedIssues: {
       'ui:title': ' ',
diff --git a/src/applications/disability-benefits/996/pages/sameOffice.js b/src/applications/disability-benefits/996/pages/sameOffice.js
index d4b2699190..2ff657139e 100644
--- a/src/applications/disability-benefits/996/pages/sameOffice.js
+++ b/src/applications/disability-benefits/996/pages/sameOffice.js
@@ -6,7 +6,7 @@ import {
 const sameOfficePage = {
   uiSchema: {
     'ui:options': {
-      ariaLabelForEditButtonOnReview: 'Edit office of review',
+      itemName: 'office of review',
     },
     'view:sameOfficeInfo': {
       'ui:title': ' ',
diff --git a/src/applications/disability-benefits/all-claims/pages/addDisabilities.js b/src/applications/disability-benefits/all-claims/pages/addDisabilities.js
index b7018d1319..7e2b3a9456 100644
--- a/src/applications/disability-benefits/all-claims/pages/addDisabilities.js
+++ b/src/applications/disability-benefits/all-claims/pages/addDisabilities.js
@@ -87,7 +87,7 @@ export const uiSchema = {
       'ui:objectViewField': ConditionReviewField,
       'ui:options': {
         itemAriaLabel: data => data.condition,
-        ariaLabelForEditButtonOnReview: 'Edit New condition',
+        itemName: 'New condition',
       },
     },
   },
diff --git a/src/applications/disability-benefits/all-claims/pages/militaryHistory.js b/src/applications/disability-benefits/all-claims/pages/militaryHistory.js
index 3b02e8bab5..2c080efe0b 100644
--- a/src/applications/disability-benefits/all-claims/pages/militaryHistory.js
+++ b/src/applications/disability-benefits/all-claims/pages/militaryHistory.js
@@ -78,7 +78,7 @@ export const uiSchema = {
         dateRange: dateRangeUISchema,
         'ui:options': {
           itemAriaLabel: data => data.serviceBranch,
-          ariaLabelForEditButtonOnReview: 'Edit Military service history',
+          itemName: 'Military service history',
         },
       },
     },
diff --git a/src/applications/pre-need/utils/helpers.js b/src/applications/pre-need/utils/helpers.js
index f4d6150dd3..997d8f6867 100644
--- a/src/applications/pre-need/utils/helpers.js
+++ b/src/applications/pre-need/utils/helpers.js
@@ -393,7 +393,7 @@ export const serviceRecordsUI = {
   items: {
     'ui:order': ['serviceBranch', '*'],
     'ui:options': {
-      ariaLabelForEditButtonOnReview: 'Service Period',
+      itemName: 'Service Period',
     },
     serviceBranch: autosuggest.uiSchema('Branch of service', null, {
       'ui:options': {
diff --git a/src/platform/forms-system/src/js/review/ObjectField.jsx b/src/platform/forms-system/src/js/review/ObjectField.jsx
index b7f76a8bf1..3ae596f3d5 100644
--- a/src/platform/forms-system/src/js/review/ObjectField.jsx
+++ b/src/platform/forms-system/src/js/review/ObjectField.jsx
@@ -146,10 +146,7 @@ class ObjectField extends React.Component {
       const itemName =
         (typeof ariaLabel === 'function' && ariaLabel(formData || {})) ||
         formData[uiOptions.itemName];
-      const editLabel =
-        (itemName && `Edit ${itemName}`) ||
-        uiOptions.ariaLabelForEditButtonOnReview ||
-        `Edit ${title}`;
+      const editLabel = (itemName && `Edit ${itemName}`) || `Edit ${title}`;
 
       const Tag = divWrapper ? 'div' : 'dl';
       const objectViewField = uiSchema?.['ui:objectViewField'];
```
 
 </details>
 
 We might not even need `itemName` in the two files where `itemAriaLabel` is there due to this PR.
 
 If we are able to remove that property, we could get the `editLabel` in `<ObjectField>` to more closely resemble what's going on in `<ArrayField>`, only with the use of `title` instead of `"Item"`:
 
 ```js
       const itemName =
        (typeof ariaLabel === 'function' && ariaLabel(formData || {})) ||
        formData[uiOptions.itemName] ||
        title;
      const editLabel = `Edit ${itemName}`;

```

Mottie committed:
Change to submit type button

Mottie committed:
Remove ariaLabelForEditButtonOnReview option

Mottie committed:
Add aria-label to "Update page" button

Mottie commented:
@bkjohnson So, in this last update... 

- Added a `type="submit` to the `ArrayField` "Update" button
- I added your suggestions, but in the `ObjectField.jsx` I ended up needing to include the options `itemName` because it was returning an undefined `itemName`

    ```diff
      const itemName =
        (typeof ariaLabel === 'function' && ariaLabel(formData || {})) ||
    +   formData[uiOptions.itemName] ||
    +   uiOptions.itemName;
    ```
- Removed all instances of `ariaLabelForEditButtonOnReview`
- And speaking of scope creep, I added an `aria-label` to the "Update page" button seen when you edit an entry on the review & submit page.

bkjohnson commented on src/platform/forms-system/src/js/components/ProgressButton.jsx:
```diff
@@ -33,6 +33,7 @@ class ProgressButton extends React.Component {
         }`}
         id={`${this.id}-continueButton`}
         onClick={this.props.onButtonClick}
+        aria-label={this.props['aria-label'] || null}
```
Can this new prop be added to the `propTypes` and be camel cased?

bkjohnson commented on src/platform/forms-system/src/js/review/ReviewCollapsibleChapter.jsx:
```diff
@@ -225,6 +228,7 @@ export default class ReviewCollapsibleChapter extends React.Component {
                         }}
                         buttonText="Update page"
                         buttonClass="usa-button-primary"
+                        aria-label={ariaLabel}
```
This seems fine as long as the prop is camel cased, just because that's the "React way"

bkjohnson reviewed (approved):
There's a minor change, but everything else seems good to go :+1:

Mottie committed:
Use camel-case props
